provider/oci/provider_test.go

```diff
@@ -0,0 +1,211 @@
+		"fingerprint": ocitesting.PrivateKeyEncryptedFingerprint,
+		"region":      "us-phoenix-1",
+	})
+}
+
+func (s *credentialsSuite) TestPassphraseHiddenAttributes(c *gc.C) {
+	envtesting.AssertProviderCredentialsAttributesHidden(c, s.provider, "httpsig", "pass-phrase")
+}
+
+func (s *credentialsSuite) TestDetectCredentialsNotFound(c *gc.C) {
+	_, err := s.provider.DetectCredentials()
+	c.Assert(errors.Cause(err), jc.Satisfies, os.IsNotExist)
+}
+
+func (s *credentialsSuite) TestDetectCredentials(c *gc.C) {
+	cfg := map[string]map[string]string{
+		"DEFAULT": map[string]string{
+			"fingerprint": ocitesting.PrivateKeyEncryptedFingerprint,
+			"pass-phrase": ocitesting.PrivateKeyPassphrase,
+			"key":         ocitesting.PrivateKeyEncrypted,
+			"region":      "us-phoenix-1",
+		},
+	}
+	s.writeOCIConfig(c, cfg)
+	creds, err := s.provider.DetectCredentials()
+	c.Assert(err, gc.IsNil)
+	c.Assert(len(creds.AuthCredentials), gc.Equals, 1)
+	c.Assert(creds.DefaultRegion, gc.Equals, "us-phoenix-1")
+}
+
+func (s *credentialsSuite) TestDetectCredentialsWrongPassphrase(c *gc.C) {
+	cfg := map[string]map[string]string{
+		"DEFAULT": map[string]string{
+			"fingerprint": ocitesting.PrivateKeyEncryptedFingerprint,
+			"pass-phrase": "bogus",
+			"key":         ocitesting.PrivateKeyEncrypted,
+			"region":      "us-phoenix-1",
+		},
+	}
+	s.writeOCIConfig(c, cfg)
+	_, err := s.provider.DetectCredentials()
+	c.Assert(err, gc.NotNil)
+}
+
+func (s *credentialsSuite) TestDetectCredentialsMultiSection(c *gc.C) {
+	cfg := map[string]map[string]string{
+		"DEFAULT": map[string]string{
+			"fingerprint": ocitesting.PrivateKeyEncryptedFingerprint,
+			"pass-phrase": ocitesting.PrivateKeyPassphrase,
+			"key":         ocitesting.PrivateKeyEncrypted,
+			"region":      "us-ashburn-1",
+		},
+		"SECONDARY": map[string]string{
+			"fingerprint": ocitesting.PrivateKeyUnencryptedFingerprint,
+			"pass-phrase": "",
+			"key":         ocitesting.PrivateKeyUnencrypted,
+			"region":      "us-phoenix-1",
+		},
+	}
+	s.writeOCIConfig(c, cfg)
+	creds, err := s.provider.DetectCredentials()
+	c.Assert(err, gc.IsNil)
+	c.Assert(len(creds.AuthCredentials), gc.Equals, 2)
+	c.Assert(creds.DefaultRegion, gc.Equals, "us-ashburn-1")
+}
+
+func (s *credentialsSuite) TestDetectCredentialsMultiSectionInvalidConfig(c *gc.C) {
+	cfg := map[string]map[string]string{
+		"DEFAULT": map[string]string{
+			"fingerprint": ocitesting.PrivateKeyEncryptedFingerprint,
+			"pass-phrase": "bogus",
+			"key":         ocitesting.PrivateKeyEncrypted,
+			"region":      "us-ashburn-1",
+		},
+		"SECONDARY": map[string]string{
+			"fingerprint": ocitesting.PrivateKeyUnencryptedFingerprint,
+			"pass-phrase": "",
+			"key":         ocitesting.PrivateKeyUnencrypted,
+			"region":      "us-phoenix-1",
+		},
+	}
+	s.writeOCIConfig(c, cfg)
+	creds, err := s.provider.DetectCredentials()
+	c.Assert(err, gc.IsNil)
+	c.Assert(len(creds.AuthCredentials), gc.Equals, 1)
+	c.Assert(creds.DefaultRegion, gc.Equals, "")
```

Comment:
I'm a bit confused. Which of the sections is invalid? I'm guessing this is meant to be testing that if I have 1 good credential, and 1 bad credential, I can still find the good one, and still make it work.
I'm guessing SECONDARY is the valid one, given it is using ...Unencrypted... and pass-phrase: "" (and DEFAULT is using "bogus".)

Is it just that we only set DefaultRegion if the DEFAULT section is valid?

Reply:
I should have left a comment. Yes, the one with the bogus password is invalid. It errors out at decrypting the key, so it continues with the rest of the sections.

The "DEFAULT" section is the one used by the ```oci``` cli when not explicitly passing in the config section to use. We have no way to decide which region is "default" without user interaction or without a "DEFAULT" section.